src/java.base/share/classes/jdk/internal/util/jar/PersistentJarFile.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2022, Azul Systems, Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.util.jar;
+
+import jdk.crac.Context;
+import jdk.crac.Resource;
+import jdk.internal.crac.Core;
+import jdk.internal.crac.JDKResource;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.jar.JarFile;
+import java.util.zip.ZipFile;
+
+public class PersistentJarFile extends JarFile implements JDKResource {
+
+    public PersistentJarFile(File file, boolean b, int openRead, Runtime.Version runtimeVersion) throws IOException {
+        super(file, b, openRead, runtimeVersion);
+        Core.getJDKContext().register(this);
+    }
+
+    @Override
+    public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+        Method zipBeforeCheckpoint = ZipFile.class.getDeclaredMethod("beforeCheckpoint");
+        @SuppressWarnings("removal")
+        Void v = AccessController.doPrivileged(new PrivilegedAction<Void>() {
+            public Void run() {
+                zipBeforeCheckpoint.setAccessible(true);
+                return null;
+            }});
```

Comment:
Something to think about that doesn't need to be addressed now:

We may want to lazily cache the accessible reflect Method in a static field to avoid looking it up for each PersistentJarFile.  The value of doing that depends somewhat on how many jars we expect here and how much we care about time spent in checkpoint hooks.

Reply:
Oh, you're right. Not sure when it will be an another chance to improve, so better be done now :) This implementation with reflection predated the SharedSecrets, which cleaner, safer, faster -- so I just used that.